src/memory_server.ml

```diff
@@ -206,15 +211,23 @@ let parse_proc_meminfo () =
 
 		) (fun () -> close_in ic)
 
+
+(* The total amount of memory addressable by this OS. If we cannot get it
+   from Xen (which may not be running if we've just installed
+   the packages and are now setting them up), then we ask the balloon driver, or
+   worst-case, /proc/meminfo. *)
 let get_total_memory () =
-	try
-		let pairs = parse_sysfs_balloon () in
-		let keys = [ _low_mem_balloon; _high_mem_balloon; _current_allocation ] in
-		let vs = List.map (fun x -> List.assoc x pairs) keys in
-		Int64.mul 1024L (List.fold_left Int64.add 0L vs)
-	with _ ->
-		error "Failed to query balloon driver; parsing /proc/meminfo instead";
-		parse_proc_meminfo ()
+	match get_total_memory_from_xen () with
```

Comment:
I think the sequencing could be clearer here.  We want to try `get_total_memory_from_xen`, then if that fails try the balloon driver, then if that fails try proc.   However the code is currently "try Xen, catching exeception and wrapping it as an option;  unwrap option and if it failed try balloon inline;  if that threw then call parse_proc_meminfo().

Could you please:
 * lift the balloon logic into its own function
 * unify the three query functions to all have the same signature - either returning options or results, or throwing exceptions
 * rewrite get_total_memory to string those three functions together in a clearer way